test/e2e/clusterctl_upgrade.go

```diff
@@ -0,0 +1,298 @@
+		// So we are creating a workload cluster that will be used as a new management cluster where to install older version of providers
+
+		clusterctl.ApplyClusterTemplateAndWait(context.TODO(), clusterctl.ApplyClusterTemplateAndWaitInput{
+			ClusterProxy: input.BootstrapClusterProxy,
+			ConfigCluster: clusterctl.ConfigClusterInput{
+				LogFolder:                filepath.Join(input.ArtifactFolder, "clusters", input.BootstrapClusterProxy.GetName()),
+				ClusterctlConfigPath:     input.ClusterctlConfigPath,
+				KubeconfigPath:           input.BootstrapClusterProxy.GetKubeconfigPath(),
+				InfrastructureProvider:   clusterctl.DefaultInfrastructureProvider,
+				Flavor:                   clusterctl.DefaultFlavor,
+				Namespace:                managementClusterNamespace.Name,
+				ClusterName:              fmt.Sprintf("%s-%s", specName, util.RandomString(6)),
+				KubernetesVersion:        input.E2EConfig.GetVariable(KubernetesVersion),
+				ControlPlaneMachineCount: pointer.Int64Ptr(1),
+				WorkerMachineCount:       pointer.Int64Ptr(1),
+			},
+			WaitForClusterIntervals:      input.E2EConfig.GetIntervals(specName, "wait-cluster"),
+			WaitForControlPlaneIntervals: input.E2EConfig.GetIntervals(specName, "wait-control-plane"),
+			WaitForMachineDeployments:    input.E2EConfig.GetIntervals(specName, "wait-worker-nodes"),
+		}, managementClusterResources)
+
+		By("Turning the workload cluster into a management cluster with older versions of providers")
+
+		// In case of the cluster id a DockerCluster, we should load controller images into the nodes.
+		// Nb. this can be achieved also by changing the DockerMachine spec, but for the time being we are using
+		// this approach because this allows to have a single source of truth for images, the e2e config
+		// Nb. the images for official version of the providers will be pulled from internet, but the latest images must be
+		// built locally and loaded into kind
+		cluster := managementClusterResources.Cluster
+		if cluster.Spec.InfrastructureRef.Kind == "DockerCluster" {
+			Expect(bootstrap.LoadImagesToKindCluster(context.TODO(), bootstrap.LoadImagesToKindClusterInput{
+				Name:   cluster.Name,
+				Images: input.E2EConfig.Images,
+			})).To(Succeed())
+		}
+
+		// Get a ClusterBroker so we can interact with the workload cluster
+		managementClusterProxy = input.BootstrapClusterProxy.GetWorkloadCluster(context.TODO(), cluster.Namespace, cluster.Name)
+
+		var binaryURL string
+		if input.E2EConfig.HasVariable("INIT_WITH_BINARY") {
+			binaryURL = input.E2EConfig.GetVariable("INIT_WITH_BINARY")
+			binaryURL = strings.ReplaceAll(binaryURL, "{OS}", runtime.GOOS)
+			binaryURL = strings.ReplaceAll(binaryURL, "{ARCH}", runtime.GOARCH)
+		}
+
+		log.Logf("downloading clusterctl binary from %s", binaryURL)
+		binaryFile := downloadToTmpFile(binaryURL)
+		defer os.Remove(binaryFile) // clean up
+
+		err := os.Chmod(binaryFile, 0744)
```

Comment:
not sure if we want to but we could move this into `downloadToTmpFile` and change the scope of the func a bit.

Reply:
I prefer to keep things separated, given that might be in future we want to download files which are not executables (e.g yaml)

Reply:
We could make the mode a parameter, but fine for me to keep it separate.